src/parser/jobs/ast/modules/Lightspeed.tsx

```diff
@@ -0,0 +1,230 @@
+
+interface LIGHTSPEED_Window {
+	start: number
+	end?: number
+
+	rotation: Array<Events['action']>
+	gcdCount: number
+	mpSavings: number
+	trailingGcdEvent?: Events['action']
+
+	deathTruncated: boolean
+}
+
+// in this module we only want to track Lightspeed windows opened by
+// the character selected for analysis. windows that clip into
+// AST Lightspeed will be marked.
+// Used DNC Technicalities as basis for this module. Rewritten from previous module for consistency purposes
+export default class Lightspeed extends Analyser {
+	static override handle = 'Lightspeed'
+	static override title = t('ast.Lightspeed.title')`Lightspeed`
+	static override displayOrder = DISPLAY_ORDER.LIGHTSPEED
+
+	@dependency private timeline!: Timeline
+	@dependency private data!: Data
+
+	private history: LIGHTSPEED_Window[] = []
+	private currentWindow: LIGHTSPEED_Window | undefined = undefined
+	private castHook?: EventHook<Events['action']>
+
+	override initialise() {
+		const lightspeedFilter = filter<Event>().status(this.data.statuses.LIGHTSPEED.id)
+
+		this.addEventHook(lightspeedFilter.type('statusApply')
+			.target(this.parser.actor.id), this.tryOpenWindow)
+		this.addEventHook(lightspeedFilter.type('statusRemove')
+			.target(this.parser.actor.id), this.tryCloseWindow)
+
+		//this.addEventHook(filter<Event>().source(this.parser.actor.id).type('action'), this.onCast)
+	}
+
+	private tryOpenWindow(event: Events['statusApply']) {
+		if (this.currentWindow === undefined) {
+			this.currentWindow = {
+				start: event.timestamp,
+				rotation: [],
+				gcdCount: 0,
+				mpSavings: 0,
+
+				deathTruncated: false,
+
+			}
+			this.castHook = this.addEventHook(
+				filter<Event>()
+					.source(this.parser.actor.id)
+					.type('action'),
+				this.onCast,
+			)
+		}
+	}
+
+	private tryCloseWindow(event: Events['statusRemove']) {
+
+		if (this.currentWindow == null) {
+			return
+		}
+
+		// Make sure all applicable statuses have fallen off before the window closes
+		this.currentWindow.end = event.timestamp
+		this.history.push(this.currentWindow)
+		this.currentWindow = undefined
+		if (this.castHook != null) {
+			this.removeEventHook(this.castHook)
+			this.castHook = undefined
+
+		}
+	}
+
+	private onCast(event: Events['action']) {
+
+		// If we don't have a window, bail
+		if (this.currentWindow == null) {
+			return
+		}
+
+		const action = this.data.getAction(event.action)
+
+		// Can't do anything else if we didn't get a valid action object
+		if (action == null) {
+			return
+		}
+
+		// Add the action to the list
+		this.currentWindow.rotation.push(event)
+		if (action.onGcd) {
+			this.currentWindow.gcdCount++
+		}
+		if (this.parser.patch.before('5.3') && action.mpCost != null) {
+			this.currentWindow.mpSavings = this.currentWindow.mpSavings + action.mpCost/2
+		}
+
+		// If we haven't recorded a trailing GCD event for this closed window, do so now
+		if (this.currentWindow.end && !this.currentWindow.trailingGcdEvent && action.onGcd) {
```

Comment:
Should really be explicitly checking these, but also I think this whole block can be reordered for simplicity:

```ts
this.currentWindow.rotation.push(event)

if (this.parser.patch.before('5.3') && action.mpCost != null) {
	this.currentWindow.mpSavings = this.currentWindow.mpSavings + action.mpCost/2
}

if (action.onGcd) {
	this.currentWindow.gcdCount++

  // If we haven't recorded a trailing GCD event for this closed window, do so now
  if (this.currentWindow.end != null && this.currentWindow.trailingGcdEvent == null) {
    this.currentWindow.trailingGcdEvent = event
  }
}
```

Reply:
updated